### ISSUE TITLE
Discussion - How we should handle MapReduce and Pipelines

### ISSUE THREAD

**Issue by chargrizzle** [potatolondon/djangae#510]
Apologies in advance for this massive wall of text

OK so i've been looking at how we should handle data processing in djangae on and off for a while now and every time I sit down and look at it I come to different conclusions.
# The situation

MapReduce is essentially a wrapper around the Pipelines library, which allows generic data processing asynchronous functionality in a similar manner to the tasks queues, only with POWER. Here are an awesome examples stolen from the Pipeline Wiki, you can see the full code at 

https://github.com/GoogleCloudPlatform/appengine-pipelines/blob/master/python/demo/main.py

and read more about Pipelines at

http://sookocheff.com/series/pipelines-api/

```
class LongCount(pipeline.Pipeline):

  def run(self, entity_kind, property_name, value):
    cursor = None
    count = 0
    while True:
      query = db.GqlQuery(
          'SELECT * FROM %s WHERE %s = :1' % (entity_kind, property_name),
          value.lower(), key_only=True, cursor=cursor)
      result = query.fetch(1000)
      count += len(result)
      if len(result) < 1000:
        return (entity_kind, property_name, value, count)
      else:
        cursor = query.cursor()

class SplitCount(pipeline.Pipeline):

  def run(self, entity_kind, property_name, *value_list):
    all_counts = []
    for value in value_list:
      stage = yield LongCount(entity_kind, property_name, value)
      all_counts.append(stage)

    yield common.Append(*all_counts)

class CountReport(pipeline.Pipeline):

  def run(self, email_address, entity_kind, property_name, *value_list):
    yield common.Log.info('UselessPipeline.coolness = %s',
                          (yield UselessPipeline()).coolness)

    split_counts = yield SplitCount(entity_kind, property_name, *value_list)
    yield common.Log.info('SplitCount result = %s', split_counts)

    with pipeline.After(split_counts):
      with pipeline.InOrder():
        yield common.Delay(seconds=1)
        yield common.Log.info('Done waiting')
        yield EmailCountReport(email_address, split_counts)

  def finalized(self):
    if not self.was_aborted:
      logging.info('All done! Found %s results', self.outputs.default.value)
```

As you can see from the control flow in CountReport, with methods like After() and InOrder() you can really do some awesome things!

However the appengine pipelines library is a blessing and a curse. To get this to properly work with djangae we need to get the Pipeline requests to execute inside the django context, which in itself should not be that big a deal, we just have to make sure we route all the /_ah/pipelines traffic, we can see this in the Pipeline class

```
  def start(self,
            idempotence_key='',
            queue_name='default',
            base_path='/_ah/pipeline',
            return_task=False,
            countdown=None,
            eta=None):
```

If we do this then we _should_ have all the pipeline requests being executed in the django context, which is nice!

This problem is when we get to the MapReduce library....

Even though the Pipelines library is used they have rolled out their own handlers so there is a status dashboard which is specific to MapReduce, and the lovely pattern we can see with Pipelines is slowly destroyed. While I can see the idea behind this it also makes everything very hard to implement without re-writing a large portion of the library. 
# Suggested reading

Start with 
https://github.com/GoogleCloudPlatform/appengine-mapreduce/blob/master/python/src/mapreduce/mapreduce_pipeline.py 

Take a look at the MapPipeline class, this looks nice and inline with the pattern we expect from pipelines, you'll see there is a MapperPipeline which is yielded from this class. The pattern starts to break down at this point with a call to control.start_map which calls the Mapper Handler directly, I recently got a PR merged that allows us to override the base_path at this point, previously this was not possible so almost total code duplication was needed to get anywhere.
# In an ideal world

So what do we even want at this point? Well for MapReduce to work nicely with djangae we need it to support an InputReader that can handle Django models, this has already mostly been written. If we can coerce the entire MapReduce library to pass through the Django context as well as the Pipelines library then in theory we can just use the vanilla API, I think this is the best option and most of my adventures in this area have been trying to make this happen.

There is a lot to cover here with this problem and I myself can't even remember all the issues until I sit down and look at my attempts, if you would like to help then please have a look at the linked code samples and add some questions/suggestions to this issue if you have any and I will continue to add issues to this ticket as a thought dump.

TL;DR - Pipelines = Task Queues on steroids, MapReduce library = Pipelines in jail

**Comment by chargrizzle** [comment]
More suggested reading

https://github.com/GoogleCloudPlatform/appengine-mapreduce/pull/56

**Comment by jacobg** [comment]
Thanks for opening this issue. I had to work around this by defining all my Pipeline classes from this base class:

```python
# A Pipeline class that re-routes handler path to what was set up by Djangae.
# It is strange that we need to do this, because not all pipelines are for
# mapreduce, but Djangae decided to set up the handlers that way so we'll
# deal with it.
class Pipeline(pipeline.Pipeline):

    def start(self, *args, **kwargs):

        # make sure base_path is set to djangae handler path
        kwargs.pop('base_path', None)
        kwargs['base_path'] = PIPELINE_BASE_PATH

        super(Pipeline, self).start(*args, **kwargs)
```

As an aside, there are a number of bugs in Djangae's mapreduce implementation. I fixed a number of them in my own code, but haven't had a chance yet to send pull requests for them.

**Comment by chargrizzle** [comment]
Yes it's nice and friendly with the Pipeline stuff but MapReduce is a bit nastier and I don't want to have to reimplement all the MR classes with a new base class. I am open to suggestions :)

**Comment by jacobg** [comment]
For some reason, the MapperPipeline doesn't seem to use the start method when yielding it. I tested this by defining one and adding a print statement, and it doesn't print. And so my MapperPipeline works fine as-is, with just the run method setting the base path. Are you seeing something different?

**Comment by chargrizzle** [comment]
https://github.com/chargrizzle/djangae/tree/pipelines

I managed to work out native pipeline and mapreduce support without any horrible wrapper, I need to add some more tests and some docs but I think I will get this merged into master by new year!!

**Comment by chargrizzle** [comment]
https://github.com/potatolondon/djangae/pull/536
